pyretailscience/analysis/product_association.py

```diff
@@ -226,83 +204,162 @@ def _calc_association(  # noqa: C901 (ignore complexity) - Excluded due to min_*
 
-        if show_progress:
-            items = tqdm(items)
-
-        cols_mask = np.zeros(sparse_matrix.shape[1], dtype=bool)
-
-        for target_item_loc, item_2 in items:
-            target_item_col_index = cols_mask.copy()
-            target_item_col_index[target_item_loc] = True
-            rows_with_target_item = sparse_matrix[:, target_item_col_index].getnnz(axis=1) == len(target_item_loc)
-
-            cooccurrences = np.array(sparse_matrix[rows_with_target_item, :].sum(axis=0)).flatten()
-            if (cooccurrences == 0).all():
-                continue
-
-            rows_with_target_item_sum = rows_with_target_item.sum()
-            coocurrence_prob = cooccurrences / row_count
-
-            target_prob = rows_with_target_item_sum / row_count
-            expected_prob = target_prob * occurence_prob
-
-            # TODO: Try to avoid constructing a pandas Dataframe
-            pa_df = pd.DataFrame(
-                {
-                    f"{value_col}_1": [item_2] * sparse_matrix.shape[1],
-                    f"{value_col}_2": unique_combo_df[value_col].cat.categories.values,
-                    "occurrences_1": rows_with_target_item_sum,
-                    "occurrences_2": occurrences,
-                    "cooccurrences": cooccurrences,
-                    "support": coocurrence_prob,
-                    "confidence": cooccurrences / rows_with_target_item_sum,
-                    "uplift": coocurrence_prob / expected_prob,
-                },
-            )
+        cooccurrences = merged_df.group_by(["item_1", "item_2"]).aggregate(cooccurrences=merged_df[group_col].nunique())
+        cooccurrences = cooccurrences.mutate(
+            support=cooccurrences.cooccurrences / total_transactions,
+        )
+        cooccurrences = cooccurrences.filter(
+            (cooccurrences.cooccurrences >= min_cooccurrences) & (cooccurrences.support >= min_support),
+        )
 
-            excl_pairs_idx = (
-                target_item_col_index
-                | (pa_df["occurrences_1"] < min_occurrences)
-                | (pa_df["occurrences_2"] < min_occurrences)
-                | (pa_df["cooccurrences"] < min_cooccurrences)
-                | (pa_df["support"] < min_support)
-                | (pa_df["confidence"] < min_confidence)
-                | (pa_df["uplift"] < min_uplift)
-            )
+        product_occurrences_1_rename = product_occurrences.rename(
+            {"item_1": value_col, "occurrences_1": "occurrences", "prob_1": "occurrence_probability"},
+        )
+        product_occurrences_2_rename = product_occurrences.rename(
+            {"item_2": value_col, "occurrences_2": "occurrences", "prob_2": "occurrence_probability"},
+        )
 
-            results.append(pa_df[~excl_pairs_idx])
+        product_pairs = ibis.join(
+            cooccurrences,
+            product_occurrences_1_rename,
+            predicates=[cooccurrences["item_1"] == product_occurrences_1_rename["item_1"]],
+        )
```

Comment:
Is there a reason you used ibis.join rather than the cooccurrences.join here and below?
```suggestion
        product_pairs = cooccurrences.join(
            product_occurrences_1_rename,
            predicates=[cooccurrences["item_1"] == product_occurrences_1_rename["item_1"]],
        )
```